shared/src/commonMain/kotlin/org/hyperskill/app/progresses/view/ProgressScreenViewStateMapper.kt

```diff
@@ -0,0 +1,89 @@
+package org.hyperskill.app.progresses.view
+
+import org.hyperskill.app.core.view.mapper.SharedDateFormatter
+import org.hyperskill.app.progresses.presentation.ProgressScreenFeature
+import org.hyperskill.app.track.domain.model.Track
+import org.hyperskill.app.track.domain.model.asLevelByProjectIdMap
+
+internal class ProgressScreenViewStateMapper(
+    private val dateFormatter: SharedDateFormatter
+) {
+    fun map(state: ProgressScreenFeature.State): ProgressScreenViewState =
+        ProgressScreenViewState(
+            trackProgressViewState = when (state.trackProgressState) {
+                ProgressScreenFeature.TrackProgressState.Idle ->
+                    ProgressScreenViewState.TrackProgressViewState.Idle
+                ProgressScreenFeature.TrackProgressState.Loading ->
+                    ProgressScreenViewState.TrackProgressViewState.Loading
+                ProgressScreenFeature.TrackProgressState.Error ->
+                    ProgressScreenViewState.TrackProgressViewState.Error
+                is ProgressScreenFeature.TrackProgressState.Content ->
+                    mapTrackProgressContent(state.trackProgressState)
+            },
+            projectProgressViewState = when (state.projectProgressState) {
+                ProgressScreenFeature.ProjectProgressState.Idle ->
+                    ProgressScreenViewState.ProjectProgressViewState.Idle
+                ProgressScreenFeature.ProjectProgressState.Loading ->
+                    ProgressScreenViewState.ProjectProgressViewState.Loading
+                ProgressScreenFeature.ProjectProgressState.Error ->
+                    ProgressScreenViewState.ProjectProgressViewState.Error
+                is ProgressScreenFeature.ProjectProgressState.Content ->
+                    when (state.trackProgressState) {
+                        ProgressScreenFeature.TrackProgressState.Idle,
+                        ProgressScreenFeature.TrackProgressState.Loading ->
+                            ProgressScreenViewState.ProjectProgressViewState.Loading
+                        ProgressScreenFeature.TrackProgressState.Error ->
+                            ProgressScreenViewState.ProjectProgressViewState.Error
+                        is ProgressScreenFeature.TrackProgressState.Content ->
+                            mapProjectProgressContent(
+                                track = state.trackProgressState.trackWithProgress.track,
+                                projectProgressContent = state.projectProgressState
+                            )
+                    }
```

Comment:
If the track progress block depends on the project progress block, then all of this separation to display content independently was unnecessary

Reply:
The **project** view state depends on **track** state to display icon of project level. When user opens progress screen track with it's progress is loading from cache, so that dependency isn't blocking, track will be already loaded when project loading from remote is finished. Separation of track and project makes sense.

Reply:
@vladkash 
We can simply not show the project-level icon until the track is loaded.

Reply:
Agree, thank you!